File: backend/app/services/encryption.py
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
import base64
import os

class EncryptionService:
    def __init__(self):
        self.salt = os.urandom(16)
    
    def derive_key(self, password: str, salt: bytes = None) -> bytes:
        if salt is None:
            salt = self.salt
            
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,
            salt=salt,
            iterations=100000,
        )
        return base64.urlsafe_b64encode(kdf.derive(password.encode()))
    
    def encrypt_message(self, message: str, key: bytes) -> str:
        f = Fernet(key)
        encrypted_message = f.encrypt(message.encode())
        return base64.b64encode(encrypted_message).decode()
    
    def decrypt_message(self, encrypted_message: str, key: bytes) -> str:
        f = Fernet(key)
        decrypted_message = f.decrypt(base64.b64decode(encrypted_message))
        return decrypted_message.decode()

File: frontend/src/services/encryption.ts
import CryptoJS from 'crypto-js';

class EncryptionService {
  private userKey: string | null = null;

  setUserKey(password: string, salt: string): void {
    // Derive key from password and salt using PBKDF2
    this.userKey = CryptoJS.PBKDF2(password, salt, {
      keySize: 256 / 32,
      iterations: 100000
    }).toString();
  }

  encryptMessage(message: string): string | null {
    if (!this.userKey) return null;
    
    try {
      const encrypted = CryptoJS.AES.encrypt(message, this.userKey).toString();
      return encrypted;
    } catch (error) {
      console.error('Encryption error:', error);
      return null;
    }
  }

  decryptMessage(encryptedMessage: string): string | null {
    if (!this.userKey) return null;
    
    try {
      const bytes = CryptoJS.AES.decrypt(encryptedMessage, this.userKey);
      const decrypted = bytes.toString(CryptoJS.enc.Utf8);
      return decrypted;
    } catch (error) {
      console.error('Decryption error:', error);
      return null;
    }
  }

  generateKeyPair(): { publicKey: string; privateKey: string } {
    // Simplified key generation for demo - in production use proper asymmetric encryption
    const privateKey = CryptoJS.lib.WordArray.random(256/8).toString();
    const publicKey = CryptoJS.SHA256(privateKey).toString();
    
    return { publicKey, privateKey };
  }

  clearKeys(): void {
    this.userKey = null;
  }
}

export const encryptionService = new EncryptionService();